src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFatTPFServerWithTranslation.java

```diff
@@ -0,0 +1,223 @@
+import se.liu.ida.hefquin.engine.queryplan.executable.impl.ClosableIntermediateResultElementSink;
+import se.liu.ida.hefquin.engine.queryproc.ExecutionContext;
+
+/**
+ * Implementation of an operator to request a (complete) TPF from a TPF server.
+ * This implementation handles pagination of the TPF; that is, it requests all
+ * the pages, one after another.
+ */
+public class ExecOpRequestTPFatTPFServerWithTranslation extends ExecOpGenericRequest<TriplePatternRequest, TPFServer>
+{
+	public ExecOpRequestTPFatTPFServerWithTranslation( final TriplePatternRequest req, final TPFServer fm ) {
+		super( req, fm );
+
+		assert fm.getVocabularyMapping() != null;
+	}
+
+	
+	@Override
+	protected void _execute(IntermediateResultElementSink sink, ExecutionContext execCxt)
+			throws ExecOpExecutionException {
+		
+		FederationAccessManager fedAccessMgr = execCxt.getFederationAccessMgr();
+		final SPARQLGraphPattern reqTranslation = fm.getVocabularyMapping().translateTriplePattern(req.getQueryPattern());
+		Iterator<SolutionMapping> res;
+		if(reqTranslation instanceof TriplePattern) {
+			res = handleTriplePattern((TriplePattern) reqTranslation, fedAccessMgr).iterator();
+		}
+		else if(reqTranslation instanceof SPARQLUnionPattern) {
+			res = handleUnionPattern(((SPARQLUnionPattern) reqTranslation), fedAccessMgr);
+		}
+		else if(reqTranslation instanceof SPARQLGroupPattern) {
+			res = handleGroupPattern(((SPARQLGroupPattern) reqTranslation), fedAccessMgr);
+		}
+		else if(reqTranslation instanceof BGP) {
+			res = handleBGP((BGP) reqTranslation, fedAccessMgr);
+		}
+		else {
+			throw new ExecOpExecutionException(reqTranslation.toString(), this);
+		}
+		
+		while(res.hasNext()) {
+			sink.send(res.next());
+		}
+		
+	}
+	
+	protected Iterable<SolutionMapping> handleTriplePattern(final TriplePattern tp, final FederationAccessManager fedAccessMgr) throws ExecOpExecutionException{
+		
+		final TriplePatternRequest newReq = new TriplePatternRequestImpl(tp);
+		final ExecOpRequestTPFatTPFServer op = new ExecOpRequestTPFatTPFServer(newReq, fm);
+		final MaterializingIntermediateResultElementSinkWithTranslation sink = new MaterializingIntermediateResultElementSinkWithTranslation(fm.getVocabularyMapping());
+		
+		TPFResponse currentPage = null;
+		while ( currentPage == null || ! op.isLastPage(currentPage) ) {
+			// create the request for the next page (which is the first page if currentPage is null)
+			final TPFRequest pageRequest = op.createPageRequest(currentPage);
+
+			// perform the page request
+			try {
+				currentPage = op.performPageRequest( pageRequest, fedAccessMgr );
+			}
+			catch ( final FederationAccessException e ) {
+				throw new ExecOpExecutionException("Issuing a page request caused an exception: " + e.toString(), this);
+			}
+			
+			// consume the matching triples retrieved via the page request
+			op.consumeMatchingTriples( currentPage.getPayload(), sink );
+
+		}	
+	
+		return sink.getMaterializedIntermediateResult();
```

Comment:
On the interest of having homogeneous return types for the handle methods, please directly return the iterator here, instead of calling the iterator method in _execute and in every other place you use handleTriplePattern.